File: tests/unit/relays/read-availability.test.ts
import { describe, expect, it } from 'vitest';
import { defaultRelaySet } from '../../../src/lib/relays/default-relays';
import {
  durableRelayReadPlan,
  readRelaysAvailable,
  unavailableRelayReadPlan,
} from '../../../src/lib/relays/read-availability';

describe('read availability', () => {
  it('distinguishes durable no-read-relay from settings unavailable', () => {
    const durableEmpty = durableRelayReadPlan([
      {
        ...defaultRelaySet,
        relays: defaultRelaySet.relays.map((relay) => ({
          ...relay,
          read: false,
        })),
      },
    ]);
    const unavailable = unavailableRelayReadPlan('opfs-owner-held', 'allowed');

    expect(durableEmpty.source).toBe('durable-empty');
    expect(durableEmpty.relays).toEqual([]);
    expect(unavailable.source).toBe('session-default-public-read');
    expect(unavailable.relays.length).toBeGreaterThan(0);
    expect(unavailable.writeAllowed).toBe(false);
    expect(readRelaysAvailable(unavailable)).toBe(true);
  });

  it('keeps policy-forbidden unavailable separate from no enabled relay', () => {
    const plan = unavailableRelayReadPlan('web-lock-unavailable', 'forbidden');

    expect(plan).toEqual({
      source: 'settings-unavailable',
      relays: [],
      diagnostic: 'Relay settings unavailable: web-lock-unavailable.',
      writeAllowed: false,
    });
  });
});

File: crates/lkjstr-app/tests/read_availability_test.rs
use lkjstr_app::read_availability::{
    EffectiveReadRelays, ReadRelaySource, SessionDefaultReadPolicy,
};

#[test]
fn durable_empty_is_not_settings_unavailable() {
    let plan = EffectiveReadRelays::from_durable_settings(Vec::new());

    assert_eq!(plan.source, ReadRelaySource::DurableEmpty);
    assert!(plan.relays.is_empty());
    assert_eq!(plan.diagnostic, None);
    assert!(plan.write_allowed);
}

#[test]
fn unavailable_with_allowed_fallback_uses_read_only_defaults() {
    let plan = EffectiveReadRelays::from_unavailable(
        "opfs-owner-held",
        SessionDefaultReadPolicy::Allowed,
        vec!["wss://relay.example".to_owned()],
    );

    assert_eq!(plan.relays, vec!["wss://relay.example"]);
    assert_eq!(
        plan.source,
        ReadRelaySource::SessionDefaultPublicRead {
            reason: "opfs-owner-held".to_owned(),
        }
    );
    assert_eq!(
        plan.diagnostic.as_deref(),
        Some(
            "Relay settings unavailable: opfs-owner-held; using session default public read relays."
        )
    );
    assert!(!plan.write_allowed);
}

#[test]
fn unavailable_with_forbidden_fallback_stays_unavailable() {
    let plan = EffectiveReadRelays::from_unavailable(
        "web-lock-unavailable",
        SessionDefaultReadPolicy::Forbidden,
        vec!["wss://relay.example".to_owned()],
    );

    assert!(plan.relays.is_empty());
    assert_eq!(
        plan.source,
        ReadRelaySource::SettingsUnavailable {
            reason: "web-lock-unavailable".to_owned(),
        }
    );
    assert_eq!(
        plan.diagnostic.as_deref(),
        Some("Relay settings unavailable: web-lock-unavailable.")
    );
    assert!(!plan.write_allowed);
}
